bridge/contracts/libraries/snapshots/SnapshotRingBuffer.sol

```diff
@@ -0,0 +1,131 @@
+// SPDX-License-Identifier: MIT-open-group
+pragma solidity ^0.8.11;
+
+import "contracts/libraries/parsers/BClaimsParserLibrary.sol";
+import "contracts/interfaces/ISnapshots.sol";
+
+struct Epoch {
+    uint32 _value;
+}
+
+struct SnapshotBuffer {
+    Snapshot[6] _array;
+}
+
+library RingBuffer {
+    function get(SnapshotBuffer storage self_, uint32 epoch_)
+        internal
+        view
+        returns (Snapshot storage)
+    {
+        return self_._array[_indexFor(self_, epoch_)];
+    }
+
+    //writes the new snapshot to the buffer
+    function set(
+        SnapshotBuffer storage self_,
+        function(uint32) returns (uint32) epochFor_,
+        Snapshot memory new_
+    ) internal returns (uint32) {
+        //get the epoch corresponding to the blocknumber
+        uint32 epoch = epochFor_(new_.blockClaims.height);
+        //gets the snapshot that was at that location of the buffer
+        Snapshot storage old = self_._array[_indexFor(self_, epoch)];
+        //checks if the new snapshot height is greater than the previous
+        require(new_.blockClaims.height > old.blockClaims.height, "invalid new blockheight");
+        unsafeSet(self_, new_, epoch);
+        return epoch;
+    }
+
+    function unsafeSet(
+        SnapshotBuffer storage self_,
+        Snapshot memory new_,
+        uint32 epoch_
+    ) internal {
+        self_._array[_indexFor(self_, epoch_)] = new_;
+    }
+
+    /**
+     * @dev calculates the congruent value for current epoch in respect to the array length
+     * for index to be replaced with most recent epoch
+     * @param epoch_ epoch_ number associated with the snapshot
+     */
+    function _indexFor(SnapshotBuffer storage self_, uint32 epoch_)
+        internal
+        view
+        returns (uint256)
+    {
+        //TODO determine if this is necessary
```

Comment:
Do we need the require below? The snapshot committed to the contract start at 1.